acvm/src/lib.rs

```diff
@@ -50,7 +48,14 @@ pub enum OpcodeResolutionError {
     #[error("unexpected opcode, expected {0}, but got {1}")]
     UnexpectedOpcode(&'static str, BlackBoxFunc),
     #[error("expected {0} inputs for function {1}, but got {2}")]
-    IncorrectNumFunctionArguments(usize, BlackBoxFunc, usize),
+    IncorrectNumFunctionArguments(usize, String, usize),
+}
+
+#[derive(Debug, PartialEq)]
+pub enum GateResolution {
+    Solved,                     // The gate is resolved
+    Stalled(OpcodeNotSolvable), // The gate is not solvable
+    InProgress,                 // The gate is not solvable but could resolved some witness
```

Comment:
Can you explain why these two variants, Stalled and InPogress are needed at this layer? It seems like an abstraction leak from the Blocksolver

Reply:
If you have only gates that are solved/unsolved you cannot determine if you are stalled or not. A gate individually cannot know if the whole process is stalled or not so you cannot confine the enum to the Blocksolver.

Reply:
We have two forms of stalling here however, a `Block` can stall and also the entire solving can stall.

The first we can catch easily within `block.solve` and return like we have in `GateResolution`.

The second is different as it's global. All the other opcode kinds can return pass/fail and still feed into this value just fine. We then just need to handle the `InProgress/Stalled` variants differently for `Block` opcodes when setting the global stalled flag.

Reply:
Making all opcodes return a `GateResolution` however creates ambiguity about "can a black box function progress but not be fully solved?" where you need to check all the potential return values.

Reply:
Indeed we check the 3 possible return values, there is no problem with that.

Reply:
Indeed I don't understand what you want to say. If someone is new to the codebase and wants to know if an Arithmetic  pass or fail, then it's easy, either the solve returns an error which means it is failing, or not meaning the pass succeeded. Furthermore, the return value indicates when the gate is solved.

Reply:
My concern isn't with seeing if a particular invocation of the solver returned a `Solved` or `Stalled` but rather that the fact that it returns a `GateResolution` implies that it is able to signal that it's made partial progress. In order to verify that the `Arithmetic` solver is "all or nothing" a human needs to verify that it never returns `GateResolution::InProgress`.

My opinion is that we should make a conscious choice about which opcodes are allowed to signal that they can partially progress to provide more information on the execution flow inside the ACVM, i.e. which opcodes return `GateResolution` vs a restricted pass/fail value.

For instance we should ask ourselves whether we want a blackbox function be able to signal partial progress? All the instances we have currently are "all or nothing" so why allow a backend to return `GateResolution::InProgress`?

Reply:
Thank you for the clarification.
For me, I don't think we should make this choice at the gate level, it is the responsibility of the solver to decide the gate status. If it can only be `solved` or `stalled`, then  you can see this in the solver.

If for example you consider arithmetic gate, a `binary solver` will be able to make some progress without solving the gate by identifying some witness as boolean. The basic `arithmetic solver` on the other is either stall or solved.

Reply:
The core of my point however is that we should codify which solvers are atomic vs non-atomic by their return value, not requiring people so scan through them for all of their potential return values.